src/dataflow/sink/avro_ocf.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fs::OpenOptions;
+
+use differential_dataflow::hashable::Hashable;
+use log::error;
+use timely::dataflow::channels::pact::Exchange;
+use timely::dataflow::operators::generic::Operator;
+use timely::dataflow::{Scope, Stream};
+
+use dataflow_types::{AvroOcfSinkConnector, Diff, Timestamp};
+use expr::GlobalId;
+use interchange::avro::{DiffPair, Encoder};
+use repr::{RelationDesc, Row};
+
+pub fn avro_ocf<G>(
+    stream: &Stream<G, (Row, Timestamp, Diff)>,
+    id: GlobalId,
+    connector: AvroOcfSinkConnector,
+    desc: RelationDesc,
+) where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let encoder = Encoder::new(desc);
+    let schema = encoder.writer_schema();
+    let sink_hash = id.hashed();
+
+    let res = OpenOptions::new().append(true).open(&connector.path);
+    let mut avro_writer = match res {
+        Ok(f) => Some(avro::Writer::new(schema.clone(), f)),
+        Err(e) => {
+            error!("creating avro ocf file writer for sink failed: {}", e);
+            None
+        }
+    };
```

Comment:
This strikes me as a little bit dangerous. If two workers write to the file, even if they both have it open in append-only mode, the result will be total nonsense. At the moment `avro::Writer::new` doesn't write to the file, but will that always be true?

The early check for whether we can create the file is definitely good in reducing the likelihood of something going wrong, but I think, even so, we might eventually want to work a bit harder to limit the number of `Writer`s that are created to one.

Anyway, I think we should merge as is for now! If this struggle is common in a lot of sinks, we can wrap it up in a utility function later.

Reply:
Well -- it's a good thing we control `avro::Writer::new` now :)

but more seriously your point is well taken and I will add comments / see if I can't address this with a refactor when I'm adding tests.

My feeling on this has evolved slightly in the past few days and I now think that any place in the code where we are trying to make control flow based choices on _which worker_ should do a thing (or really anything that looks like a lock/semaphore if you squint hard enough) is kind of a giant antipattern with differential that's way too easy to get wrong and I want to try to avoid as hard as I can

Reply:
also TIL -- I had mistakenly thought that O_APPEND meant writes were always atomic but actually it depends on filesystem and even then its only usually up to page size (4 kb)

I guess it was too good to be true

Reply:
Yeah and even if they are atomic, having multiple Avro writers means multiple things will try to write out the file header—you won't end up with a valid Avro file. 

> My feeling on this has evolved slightly in the past few days and I now think that any place in the code where we are trying to make control flow based choices on which worker should do a thing (or really anything that looks like a lock/semaphore if you squint hard enough) is kind of a giant antipattern with differential that's way too easy to get wrong and I want to try to avoid as hard as I can

I'd advocate for a slightly modified version of this. Every worker needs to build an identical _dataflow_, but it's ok if _within_ an individual operator the behavior differs by worker. This distinction is subtle, and not at all easy to discern when you first wade into differential/timely code. But I think this is just a necessary fact of dealing with inputs and outputs that must be ordered: necessarily only one worker can be responsible for consuming or producing that order. 

I think, though, we could wrap up this pattern in a utility function called `sink_single_worker`, or something like that, where it is impossible to misuse the API, and the code with the conditional can be tight and well-vetted to ensure that it builds an identical dataflow, and only the implementation differs between workers.